> # **NUZAIF NAVEED - 21418890**

# Reading the Dataset
- Step 1: This will be able to read the file in a Spark Dataframe.
- Step 2: This will create a Temporary SQL View, which allows us to run SQL queries directly on the dataframe as if it were a table.
- Step 3: This will query the data in a new Spark DataFrame, and rename the headers in order to be used in the code more efficiently.
- Step 4: Displaying the data.

In [0]:


# Step 1
AirlineDF = spark.read.format("csv")\
   .option("inferSchema", "true")\
   .option("header", "true")\
   .load("/FileStore/tables/Homework_03-03-2025/Air_Traffic_Passenger_Statistics1999toOct2023.csv")


# Step 2
AirlineDF.createOrReplaceTempView("Airline_Data_View")

# Step 3
AirlineDF_4_SQL = spark.sql(""" SELECT `Activity Period` ActivityPeriod,  \
                           `Activity Period Start Date` ActivityPeriodStartDate, \
                           `Operating Airline` OperatingAirline, \
                           `Operating Airline IATA Code`	OperatingAirlineIATACode, \
                           `Published Airline` PublishedAirline, \
                           `Published Airline IATA Code` PublishedAirlineIATACode, \
                           `GEO Summary` GEOSummary, \
                           `GEO Region` GEORegion, \
                           `Activity Type Code` ActivityTypeCode, \
                           `Price Category Code` PriceCategoryCode, \
                           Terminal, \
                           `Boarding Area` BoardingArea, \
                           `Passenger Count`	PassengerCount 
                           FROM Airline_Data_View """)

# Step 4
print()
print("DataFrame View")
display(AirlineDF_4_SQL.limit(10))




DataFrame View


ActivityPeriod ActivityPeriodStartDate OperatingAirline OperatingAirlineIATACode PublishedAirline PublishedAirlineIATACode GEOSummary GEORegion ActivityTypeCode PriceCategoryCode Terminal BoardingArea PassengerCount 199907 1999-07-01 ATA Airlines TZ ATA Airlines TZ Domestic US Deplaned Low Fare Terminal 1 B 31432 199907 1999-07-01 ATA Airlines TZ ATA Airlines TZ Domestic US Enplaned Low Fare Terminal 1 B 31353 199907 1999-07-01 ATA Airlines TZ ATA Airlines TZ Domestic US Thru / Transit Low Fare Terminal 1 B 2518 199907 1999-07-01 Aeroflot Russian International Airlines null Aeroflot Russian International Airlines null International Europe Deplaned Other Terminal 2 D 1324 199907 1999-07-01 Aeroflot Russian International Airlines null Aeroflot Russian International Airlines null International Europe Enplaned Other Terminal 2 D 1198 199907 1999-07-01 Air Canada AC Air Canada AC International Canada Deplaned Other Terminal 1 B 24124 199907 1999-07-01 Air Canada AC Air Canada AC International Canada Enplaned Other Terminal 1 B 23613 199907 1999-07-01 Air China CA Air China CA International Asia Deplaned Other Terminal 2 D 4983 199907 1999-07-01 Air China CA Air China CA International Asia Enplaned Other Terminal 2 D 4604 199907 1999-07-01 Air Europe PE Air Europe PE International Europe Deplaned Other Terminal 2 D 205

# Question 1: Pareto Analysis
### - Number of Passengers travelled yearly inside USA
### - Month-Year that had the most number of Passenger trips 
### - Month-Year that had the least number of Passenger trips 

In [0]:

# Temporary View of Spark Dataframe for Querying. 
AirlineDF_4_SQL.createOrReplaceTempView("AirlineDF_4_SQL_View")


print()
print('Pareto Analysis of Airlines in Geographical Region of the Unites States of America')

PassengersDF = spark.sql(""" SELECT ActivityPeriod, \
                           SUM( PassengerCount ) AS SumPassengerCount \
                           FROM AirlineDF_4_SQL_View \
                           WHERE GEORegion = 'US' AND ActivityTypeCode = 'Enplaned' \
                           GROUP BY ActivityPeriod \
                           ORDER BY SumPassengerCount DESC
                           """)

display(PassengersDF.limit(10))

print()
print('Most number of Trips')

MostTripsPeriodDF = spark.sql("""
                              SELECT ActivityPeriod, 
                                     SUM(PassengerCount) AS SumPassengerCount
                              FROM AirlineDF_4_SQL_View
                              WHERE GEORegion = 'US'
                                   AND ActivityTypeCode = 'Enplaned'
                              GROUP BY ActivityPeriod
                              ORDER BY SumPassengerCount DESC
                              LIMIT 1
                              """)

display(MostTripsPeriodDF)

print()
print('Least number of Trips')

LeastTripsPeriodDF = spark.sql("""
                              SELECT ActivityPeriod, 
                                     SUM(PassengerCount) AS SumPassengerCount
                              FROM AirlineDF_4_SQL_View
                              WHERE GEORegion = 'US'
                                   AND ActivityTypeCode = 'Enplaned'
                              GROUP BY ActivityPeriod
                              ORDER BY SumPassengerCount ASC
                              LIMIT 1
                              """)

display(LeastTripsPeriodDF)



Pareto Analysis of Airlines in Geographical Region of the Unites States of America


ActivityPeriod SumPassengerCount 201908 2124944 201807 2097945 201808 2065900 201708 2058661 201806 2046234 201907 2037296 201707 2029964 201906 2016628 201706 1931560 201607 1928415 201608 1926679 201606 1893727 201805 1892932 201710 1876452 201810 1831214 201712 1824399 201905 1820053 201507 1815878 201508 1808823 201709 1808286 201803 1804536 201804 1804175 201711 1803432 201910 1796957 201705 1762585 201506 1760459 201610 1759623 201912 1753492 201605 1747303 201903 1744583 201408 1743348 201510 1734471 201609 1733944 201904 1727371 202307 1727174 201812 1726226 201809 1720818 201208 1705492 201407 1705040 201811 1699995 202306 1695181 201704 1694127 201703 1690268 201406 1677202 201308 1674690 201611 1674626 201612 1671449 201505 1666018 202308 1665705 201207 1655223 201911 1653046 201509 1631556 201512 1622065 201604 1619586 201603 1619266 201306 1615320 199907 1612097 201206 1610250 201410 1606936 201405 1602808 201307 1601990 199908 1601672 201511 1595611 201909 1591838 201503 1579309 201504 1576925 200006 1569313 200007 1564954 202305 1564568 200008 1551410 202206 1545758 201305 1529544 201310 1527896 201801 1527281 201404 1525445 201412 1525051 201312 1522640 201409 1518479 201210 1515249 202207 1510404 201107 1508056 202210 1501270 201108 1499109 201403 1491577 202001 1488874 202208 1483142 201209 1482529 201802 1478679 201205 1478095 201106 1468908 201901 1465497 201309 1463077 201411 1458968 202209 1446118 200108 1443616 202002 1439955 202303 1436756 201303 1435594 202304 1435437 201008 1427328 201701 1421306 201304 1416294 199910 1413969 201007 1409698 202205 1409528 201110 1407699 201006 1401267 200908 1400696 201204 1399368 201311 1397265 201211 1396383 200907 1396344 201109 1395349 201601 1394304 200005 1392674 202211 1387182 201203 1386529 201902 1384813 202212 1382565 201212 1371387 201112 1370770 200808 1364930 201602 1362401 201105 1359517 200004 1358890 199909 1357451 201010 1355953 201501 1351971 202204 1347394 200807 1347210 200106 1342959 201111 1338638 200906 1331989 200003 1322169 200107 1319624 201702 1319309 200010 1316767 200806 1314679 200011 1312710 201401 1306268 200009 1303318 201009 1299521 201005 1293854 202203 1292133 199911 1288641 200708 1282108 200910 1278298 199912 1271277 200012 1261472 201502 1257293 201011 1255274 200805 1252251 201104 1246822 200909 1244710 201012 1244293 200707 1241966 200905 1239919 200706 1235679 201004 1234768 201201 1228773 201103 1227591 200105 1224540 201301 1221356 201003 1213342 200208 1209153 200407 1206753 200912 1205993 202112 1205517 202111 1201975 200408 1199605 200810 1199301 201402 1196235 200104 1193681 200103 1186188 201202 1186008 200507 1185756 200207 1185176 200607 1181556 200911 1180019 200710 1179361 200508 1179285 202107 1178760 202110 1177345 200803 1175632 201302 1171607 200406 1169887 202302 1169702 200904 1165732 200606 1165103 202301 1164679 200506 1164559 200804 1154922 200608 1154879 200809 1146897 200903 1137520 200206 1135347 200705 1130868 200812 1129560 202108 1121854 200308 1121344 200711 1108660 200709 1107711 200307 1103294 200002 1092967 200712 1088225 200001 1086822 200704 1073778 201101 1072376 200811 1071866 200410 1067002 200101 1059805 200505 1057082 200610 1054853 200703 1045864 202109 1044634 201001 1042552 200510 1040518 200605 1039657 200405 1039291 200306 1033527 200604 1029324 200509 1026326 200205 1022719 200404 1018228 200503 1016133 200609 1011306 200409 1009685 200612 1006997 202106 1006372 201102 1004488 200512 1003748 200412 1000087 200603 1000017 200102 999235 200511 995862 200611 994909 200212 987729 200203 974472 200504 970081 200403 968885 201002 968026 200802 966935 200901 965559 200312 961487 200411 959057 200210 953712 202202 948859 200310 947182 200204 946145 200801 935024 200110 912932 200209 911398 200112 910961 200311 904466 200902 903367 200111 902858 200305 886937 200309 876717 200701 876459 200303 871208 200601 862833 200211 861563 200501 858688 200304 855694 200702 841826 200201 816858 202201 816


Most number of Trips


ActivityPeriod SumPassengerCount 201908 2124944


Least number of Trips


ActivityPeriod SumPassengerCount 202004 60856

# Note: Airlines carrying more/less number of passengers likely had the most/least number of flights.

## The period that had the most number of passenger trips was August 2019 with 2,124,944 passengers. 

## The period that had the least number of passenger trips was April 2020 with 60,856 passengers. 

# Question 2: Number of Passenger Trend
### - Line Chart in MS Excel

![Passenger Trips Trend](/files/tables/Homework_1_Graph.png)


### The number of passengers has been increasing overtime, considering a significant dip in the number of passengers during covid in 2020.


# Question 3: United Airlines
## - The number of Operating Airlines published under United Airlines
## - Most trips of an operating airline published under United Airlines (Note: Operating Airlines that carried the most passengers likely had the most trips)

In [0]:
print()
print('The number of Operating Airlines published under United Airlines')

OperatingAirlinesCountDF = spark.sql("""
    SELECT COUNT(DISTINCT OperatingAirline) AS NumOperatingAirlines
    FROM AirlineDF_4_SQL_View
    WHERE PublishedAirline = 'United Airlines'
        AND GEORegion = 'US'
            AND ActivityTypeCode = 'Enplaned'
""")

display(OperatingAirlinesCountDF)

print()
print('Most trips of an operating airline published under United Airlines')

MostTripsAirlineDF = spark.sql("""
    SELECT OperatingAirline, 
           SUM(PassengerCount) AS TotalPassengers
    FROM AirlineDF_4_SQL_View
    WHERE PublishedAirline = 'United Airlines'
        AND GEORegion = 'US'
            AND ActivityTypeCode = 'Enplaned'
    GROUP BY OperatingAirline
    ORDER BY TotalPassengers DESC
    LIMIT 1
""")

display(MostTripsAirlineDF)




The number of Operating Airlines published under United Airlines


NumOperatingAirlines 2


Most trips of an operating airline published under United Airlines


OperatingAirline TotalPassengers United Airlines 83016180

# Question 4: Pareto Analysis of Low Fare Airlines

In [0]:
print()
print('Top 5 Airlines with low fare by number of passengers')

Top5LowFareAirlinesDF = spark.sql("""
    SELECT OperatingAirline, 
           SUM(PassengerCount) AS TotalPassengers
    FROM AirlineDF_4_SQL_View
    WHERE PriceCategoryCode = 'Low Fare'
        AND GEORegion = 'US'
            AND ActivityTypeCode = 'Enplaned'
    GROUP BY OperatingAirline
    ORDER BY TotalPassengers DESC
    LIMIT 5
""")

display(Top5LowFareAirlinesDF)


Top 5 Airlines with low fare by number of passengers


OperatingAirline TotalPassengers Southwest Airlines 22608342 Virgin America 18584980 JetBlue Airways 8741138 Frontier Airlines 5027530 US Airways 3465570